// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/UploadFilePickerTab/UploadFilePickerTab.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import * as React from 'react';
import { GeneralHelper } from '../../../../helpers/GeneralHelper';
import { PrimaryButton, DefaultButton } from 'office-ui-fabric-react/lib/components/Button';
import { css } from '@uifabric/utilities/lib/css';
import * as strings from 'PropertyControlStrings';
import styles from './UploadFilePickerTab.module.scss';
var UploadFilePickerTab = /** @class */ (function (_super) {
    __extends(UploadFilePickerTab, _super);
    function UploadFilePickerTab(props) {
        var _this = _super.call(this, props) || this;
        /**
         * Gets called when a file is uploaded
         */
        _this._handleFileUpload = function (event) {
            if (!event.target.files || event.target.files.length < 1) {
                return;
            }
            // Get the files that were uploaded
            var files = event.target.files;
            // Grab the first file -- there should always only be one
            var file = files[0];
            var filePickerResult = {
                fileAbsoluteUrl: null,
                fileName: file.name,
                fileNameWithoutExtension: GeneralHelper.getFileNameWithoutExtension(file.name),
                downloadFileContent: function () { return Promise.resolve(file); }
            };
            if (GeneralHelper.isImage(file.name)) {
                // Convert to base64 image
                var reader_1 = new FileReader();
                reader_1.readAsDataURL(file);
                reader_1.onload = function () {
                    _this.setState({
                        filePreview: reader_1.result
                    });
                };
            }
            _this.setState({
                filePickerResult: filePickerResult,
                filePreview: undefined
            });
        };
        /**
         * Saves base64 encoded image back to property pane file picker
         */
        _this._handleSave = function () {
            _this.props.onSave(_this.state.filePickerResult);
        };
        /**
         * Closes tab without saving
         */
        _this._handleClose = function () {
            _this.props.onClose();
        };
        _this.state = {
            filePickerResult: undefined
        };
        return _this;
    }
    UploadFilePickerTab.prototype.render = function () {
        var _this = this;
        var _a = this.state, filePickerResult = _a.filePickerResult, filePreview = _a.filePreview;
        var fileName = filePickerResult ? filePickerResult.fileName : null;
        var acceptedFilesExtensions = this.props.accepts ? this.props.accepts.join(",") : null;
        return (React.createElement("div", { className: styles.tabContainer },
            React.createElement("div", { className: styles.tabHeaderContainer },
                React.createElement("h2", { className: styles.tabHeader }, strings.UploadFileHeader)),
            React.createElement("div", { className: css(styles.tab, styles.tabOffset) },
                React.createElement("input", { className: styles.localTabInput, type: "file", id: "fileInput", accept: acceptedFilesExtensions, multiple: false, onChange: function (event) { return _this._handleFileUpload(event); } }),
                fileName && filePreview &&
                    /** Display image preview */
                    React.createElement("div", { className: styles.localTabSinglePreview },
                        React.createElement("img", { className: styles.localTabSinglePreviewImage, src: filePreview, alt: filePickerResult.fileName }),
                        React.createElement("span", null, fileName)),
                React.createElement("div", null,
                    React.createElement("label", { className: styles.localTabFilename }, (!filePreview && fileName ? fileName : ""))),
                React.createElement("label", { className: styles.localTabLabel, htmlFor: "fileInput" }, (fileName ? strings.ChangeFileLinkLabel : strings.ChooseFileLinkLabel))),
            React.createElement("div", { className: styles.actionButtonsContainer },
                React.createElement("div", { className: styles.actionButtons },
                    React.createElement(PrimaryButton, { disabled: !filePickerResult, onClick: function () { return _this._handleSave(); }, className: styles.actionButton }, strings.AddFileButtonLabel),
                    React.createElement(DefaultButton, { onClick: function () { return _this._handleClose(); }, className: styles.actionButton }, strings.CancelButtonLabel)))));
    };
    return UploadFilePickerTab;
}(React.Component));
export default UploadFilePickerTab;
