// @pnp/sp/navigation/types.js
import { __awaiter, __decorate } from "tslib";
import { _SharePointQueryableInstance, _SharePointQueryableCollection, _SharePointQueryable, spInvokableFactory, deleteable, } from "../sharepointqueryable.js";
import { assign } from "@pnp/common";
import { metadata } from "../utils/metadata.js";
import { body, headers } from "@pnp/odata";
import { defaultPath } from "../decorators.js";
import { spPost } from "../operations.js";
import { tag } from "../telemetry.js";
/**
 * Represents a collection of navigation nodes
 *
 */
export class _NavigationNodes extends _SharePointQueryableCollection {
    /**
     * Gets a navigation node by id
     *
     * @param id The id of the node
     */
    getById(id) {
        return tag.configure(NavigationNode(this).concat(`(${id})`), "nns.getById");
    }
    /**
     * Adds a new node to the collection
     *
     * @param title Display name of the node
     * @param url The url of the node
     * @param visible If true the node is visible, otherwise it is hidden (default: true)
     */
    add(title, url, visible = true) {
        return __awaiter(this, void 0, void 0, function* () {
            const postBody = body(assign(metadata("SP.NavigationNode"), {
                IsVisible: visible,
                Title: title,
                Url: url,
            }));
            const data = yield spPost(this.clone(NavigationNodes, null), postBody);
            return {
                data,
                node: this.getById(data.Id),
            };
        });
    }
    /**
     * Moves a node to be after another node in the navigation
     *
     * @param nodeId Id of the node to move
     * @param previousNodeId Id of the node after which we move the node specified by nodeId
     */
    moveAfter(nodeId, previousNodeId) {
        const postBody = body({
            nodeId: nodeId,
            previousNodeId: previousNodeId,
        });
        return spPost(this.clone(NavigationNodes, "MoveAfter"), postBody);
    }
}
__decorate([
    tag("nns.add")
], _NavigationNodes.prototype, "add", null);
__decorate([
    tag("nns.moveAfter")
], _NavigationNodes.prototype, "moveAfter", null);
export const NavigationNodes = spInvokableFactory(_NavigationNodes);
/**
 * Represents an instance of a navigation node
 *
 */
export class _NavigationNode extends _SharePointQueryableInstance {
    constructor() {
        super(...arguments);
        this.delete = deleteable("nn");
    }
    /**
     * Represents the child nodes of this node
     */
    get children() {
        return tag.configure(NavigationNodes(this, "children"), "nn.children");
    }
    /**
     * Updates this node
     *
     * @param properties Properties used to update this node
     */
    update(properties) {
        return __awaiter(this, void 0, void 0, function* () {
            const postBody = body(assign(metadata("SP.NavigationNode"), properties), headers({ "X-HTTP-Method": "MERGE" }));
            const data = yield spPost(this, postBody);
            return {
                data,
                node: this,
            };
        });
    }
}
__decorate([
    tag("nn.update")
], _NavigationNode.prototype, "update", null);
export const NavigationNode = spInvokableFactory(_NavigationNode);
/**
 * Exposes the navigation components
 *
 */
let _Navigation = class _Navigation extends _SharePointQueryable {
    /**
     * Gets the quicklaunch navigation nodes for the current context
     *
     */
    get quicklaunch() {
        return tag.configure(NavigationNodes(this, "quicklaunch"), "n.quicklaunch");
    }
    /**
     * Gets the top bar navigation nodes for the current context
     *
     */
    get topNavigationBar() {
        return tag.configure(NavigationNodes(this, "topnavigationbar"), "n.topnavigationbar");
    }
};
_Navigation = __decorate([
    defaultPath("navigation")
], _Navigation);
export { _Navigation };
export const Navigation = spInvokableFactory(_Navigation);
/**
 * Represents the top level navigation service
 */
export class _NavigationService extends _SharePointQueryable {
    constructor(path = null) {
        super("_api/navigation", path);
    }
    /**
     * The MenuState service operation returns a Menu-State (dump) of a SiteMapProvider on a site.
     *
     * @param menuNodeKey MenuNode.Key of the start node within the SiteMapProvider If no key is provided the SiteMapProvider.RootNode will be the root of the menu state.
     * @param depth Depth of the dump. If no value is provided a dump with the depth of 10 is returned
     * @param mapProviderName The name identifying the SiteMapProvider to be used
     * @param customProperties comma seperated list of custom properties to be returned.
     */
    getMenuState(menuNodeKey = null, depth = 10, mapProviderName = null, customProperties = null) {
        return spPost(NavigationService("MenuState"), body({
            customProperties,
            depth,
            mapProviderName,
            menuNodeKey,
        }));
    }
    /**
     * Tries to get a SiteMapNode.Key for a given URL within a site collection.
     *
     * @param currentUrl A url representing the SiteMapNode
     * @param mapProviderName The name identifying the SiteMapProvider to be used
     */
    getMenuNodeKey(currentUrl, mapProviderName = null) {
        return spPost(NavigationService("MenuNodeKey"), body({
            currentUrl,
            mapProviderName,
        }));
    }
}
__decorate([
    tag("ns.getMenuState")
], _NavigationService.prototype, "getMenuState", null);
__decorate([
    tag("ns.getMenuNodeKey")
], _NavigationService.prototype, "getMenuNodeKey", null);
export const NavigationService = (path) => new _NavigationService(path);
