// @fluentui/react-northstar/dist/commonjs/themes/teams/components/TextArea/textAreaVariables.js
"use strict";

exports.__esModule = true;
exports.textAreaVariables = void 0;

var _utils = require("../../../../utils");

var textAreaVariables = function textAreaVariables(siteVars) {
  return {
    margin: '0',
    padding: (0, _utils.pxToRem)(7) + " " + (0, _utils.pxToRem)(12),
    borderColor: 'transparent',
    borderRadius: (0, _utils.pxToRem)(3) + " " + (0, _utils.pxToRem)(3) + " " + (0, _utils.pxToRem)(2) + " " + (0, _utils.pxToRem)(2),
    borderWidth: "0 0 " + (0, _utils.pxToRem)(2) + " 0",
    backgroundColor: siteVars.colorScheme.default.background2,
    invertedBackgroundColor: siteVars.colorScheme.default.background,
    placeholderColor: siteVars.colorScheme.default.foreground1,
    disabledBackgroundColor: siteVars.colorScheme.default.backgroundColorDisabled,
    disabledColor: siteVars.colorScheme.default.colorDisabled,
    fontColor: siteVars.colorScheme.default.foreground,
    fontSize: siteVars.fontSizes.medium,
    borderColorFocus: "transparent transparent " + siteVars.colorScheme.brand.borderFocus1 + " transparent",
    height: 'auto'
  };
};

exports.textAreaVariables = textAreaVariables;
