// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/StockImagesTab/StockImages.module.scss.js
/* tslint:disable */
require("./StockImages.module.css");
var styles = {
    filePicker: 'filePicker_fa0f3bef',
    tabsContainer: 'tabsContainer_fa0f3bef',
    tabContainer: 'tabContainer_fa0f3bef',
    nav: 'nav_fa0f3bef',
    tabHeaderContainer: 'tabHeaderContainer_fa0f3bef',
    breadcrumbNav: 'breadcrumbNav_fa0f3bef',
    breadcrumbNavItem: 'breadcrumbNavItem_fa0f3bef',
    tab: 'tab_fa0f3bef',
    tabOffset: 'tabOffset_fa0f3bef',
    tabFiles: 'tabFiles_fa0f3bef',
    scrollablePaneWrapper: 'scrollablePaneWrapper_fa0f3bef',
    actionButtonsContainer: 'actionButtonsContainer_fa0f3bef',
    actionButtons: 'actionButtons_fa0f3bef',
    actionButton: 'actionButton_fa0f3bef',
    tabHeader: 'tabHeader_fa0f3bef',
    selectedTab: 'selectedTab_fa0f3bef',
    header: 'header_fa0f3bef',
    xlg: 'xlg_fa0f3bef',
    xxlg: 'xxlg_fa0f3bef',
    xxxlg: 'xxxlg_fa0f3bef',
    xxxxlg: 'xxxxlg_fa0f3bef',
    itemPickerTopBar: 'itemPickerTopBar_fa0f3bef',
    commandBarNoChevron: 'commandBarNoChevron_fa0f3bef',
    fileItem: 'fileItem_fa0f3bef',
    folderItem: 'folderItem_fa0f3bef',
    fileTypeIcon: 'fileTypeIcon_fa0f3bef',
    fileTypeIconIcon: 'fileTypeIconIcon_fa0f3bef',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_fa0f3bef',
    iconColumnHeader: 'iconColumnHeader_fa0f3bef',
    fileRow: 'fileRow_fa0f3bef',
    folderRow: 'folderRow_fa0f3bef',
    emptyFolder: 'emptyFolder_fa0f3bef',
    emptyFolderImage: 'emptyFolderImage_fa0f3bef',
    emptyFolderImageTag: 'emptyFolderImageTag_fa0f3bef',
    emptyFolderTitle: 'emptyFolderTitle_fa0f3bef',
    emptyFolderSubText: 'emptyFolderSubText_fa0f3bef',
    emptyFolderMobile: 'emptyFolderMobile_fa0f3bef',
    emptyFolderPc: 'emptyFolderPc_fa0f3bef',
    emptyFolderIcon: 'emptyFolderIcon_fa0f3bef',
    stockImagesPickerContainer: 'stockImagesPickerContainer_fa0f3bef',
    stockImagesPicker: 'stockImagesPicker_fa0f3bef'
};
export default styles;
/* tslint:enable */
