// @fluentui/react-northstar/dist/dts/src/themes/teams/siteVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../utils");
var colors_1 = require("./colors");
//
// COLORS
//
var categoryColors_1 = require("./categoryColors");
exports.categoryColors = categoryColors_1.categoryColors;
exports.categoryColorScheme = categoryColors_1.categoryColorScheme;
var colors_2 = require("./colors");
exports.colors = colors_2.colors;
exports.contextualColors = colors_2.contextualColors;
exports.naturalColors = colors_2.naturalColors;
exports.primitiveColors = colors_2.primitiveColors;
exports.colorScheme = colors_2.colorScheme;
exports.transparentColors = colors_2.transparentColors;
//
// BORDER STYLES
//
exports.borderWidth = '1px';
exports.borderRadius = '2px';
exports.focusInnerBorderColor = colors_1.colors.white;
exports.focusOuterBorderColor = colors_1.colors.black;
//
// SHADOW LEVELS
//
exports.shadowLevel1 = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .1)';
exports.shadowLevel2 = '0 .4rem .7rem -.1rem rgba(0, 0, 0, .1)';
exports.shadowLevel3 = '0 .8rem 1rem -.2rem rgba(0, 0, 0, .1)';
exports.shadowLevel4 = '0 1.6rem 1.8rem -.4rem rgba(0, 0, 0, .1)';
exports.shadowLevel1Dark = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .25)';
//
// FONT SIZES
//
exports.fontSizes = {
    smaller: utils_1.pxToRem(10),
    small: utils_1.pxToRem(12),
    medium: utils_1.pxToRem(14),
    large: utils_1.pxToRem(18),
    larger: utils_1.pxToRem(24),
};
//
// FONT WEIGHTS
//
exports.fontWeightLight = 200;
exports.fontWeightSemilight = 300;
exports.fontWeightRegular = 400;
exports.fontWeightSemibold = 600;
exports.fontWeightBold = 700;
//
// LINE HEIGHTS
//
exports.lineHeightDefault = 1;
exports.lineHeightSmaller = 1.2;
exports.lineHeightSmall = 1.3333;
exports.lineHeightMedium = 1.4286;
exports.lineHeightLarge = 1.3333;
exports.lineHeightLarger = 1.3333;
//
// Z-INDEX
// Used to maintain proper stack order of components
//
exports.zIndexes = {
    background: 0,
    foreground: 1,
    menuItem: 2,
    overlay: 1000,
    overlayPriority: 1001,
    debug: 999999999,
};
//
// SEMANTIC ASSIGNMENTS
//
exports.bodyPadding = 0;
exports.bodyMargin = 0;
exports.bodyFontFamily = '"Segoe UI", system-ui, "Apple Color Emoji", "Segoe UI Emoji", sans-serif';
exports.bodyFontSize = exports.fontSizes.medium;
exports.bodyBackground = colors_1.colors.white;
exports.bodyColor = colors_1.colors.grey[750];
exports.bodyLineHeight = exports.lineHeightMedium;
