// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Label/labelVariables.js
"use strict";

exports.__esModule = true;
exports.labelVariables = exports.labelColorAreas = void 0;

var _utils = require("../../../../utils");

var _colorUtils = require("../../../colorUtils");

var labelColorAreas = (0, _utils.stringLiteralsArray)('foreground', 'background');
exports.labelColorAreas = labelColorAreas;

var labelVariables = function labelVariables(siteVars) {
  var colorScheme = (0, _colorUtils.extendColorScheme)(siteVars.colorScheme, {
    default: {
      background: 'rgba(0, 0, 0, 0.6)',
      foreground: 'rgb(232, 232, 232)'
    },
    brand: {
      background: siteVars.colorScheme.brand.foreground4
    },
    red: {
      background: siteVars.colorScheme.red.foreground1
    }
  });
  return {
    colorScheme: (0, _colorUtils.pickValuesFromColorScheme)(colorScheme, labelColorAreas),
    circularRadius: (0, _utils.pxToRem)(9999),
    iconSize: (0, _utils.pxToRem)(16),
    padding: "0 " + (0, _utils.pxToRem)(4) + " 0 " + (0, _utils.pxToRem)(4),
    startPaddingLeft: '0px',
    endPaddingRight: '0px',
    height: (0, _utils.pxToRem)(20)
  };
};

exports.labelVariables = labelVariables;
