// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Label/labelStyles.js
"use strict";

exports.__esModule = true;
exports.labelStyles = void 0;

var _utils = require("../../../../utils");

var _colors = require("../../colors");

var labelStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    var colors = (0, _colors.getColorScheme)(v.colorScheme, p.color);
    return Object.assign({
      display: 'inline-flex',
      alignItems: 'center',
      overflow: 'hidden',
      height: v.height,
      lineHeight: v.height,
      color: colors.background,
      backgroundColor: colors.foreground,
      fontSize: (0, _utils.pxToRem)(14),
      borderRadius: (0, _utils.pxToRem)(3),
      padding: v.padding
    }, p.hasImage && (p.imagePosition === 'start' ? {
      paddingLeft: v.startPaddingLeft
    } : {
      paddingRight: v.endPaddingRight
    }), p.circular && {
      borderRadius: v.circularRadius
    });
  },
  content: function content(_ref2) {
    var p = _ref2.props,
        v = _ref2.variables;
    var hasStartElement = p.hasImage && p.imagePosition === 'start' || p.hasIcon && p.iconPosition === 'start';
    var hasEndElement = p.hasImage && p.imagePosition === 'end' || p.hasIcon && p.iconPosition === 'end';
    return Object.assign({}, hasStartElement && {
      marginLeft: (0, _utils.pxToRem)(3)
    }, hasEndElement && {
      marginRight: (0, _utils.pxToRem)(3)
    });
  },
  image: function image(_ref3) {
    var v = _ref3.variables;
    return {
      height: v.height,
      width: v.height
    };
  },
  icon: function icon(_ref4) {
    var p = _ref4.props,
        v = _ref4.variables;
    return Object.assign({
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center',
      width: v.iconSize,
      height: v.iconSize,
      '& > :first-child': {
        height: '100%',
        width: '100%',
        '& svg': {
          height: '100%',
          width: '100%'
        }
      }
    }, p.hasActionableIcon && {
      cursor: 'pointer'
    });
  }
};
exports.labelStyles = labelStyles;
