// @fluentui/accessibility/dist/dts/src/behaviors/Accordion/accordionBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var keyboard_key_1 = require("@fluentui/keyboard-key");
/**
 * @specification
 * Adds attribute 'role=presentation' to 'root' slot.
 * Triggers 'moveNext' action with 'ArrowDown' on 'root'.
 * Triggers 'movePrevious' action with 'ArrowUp' on 'root'.
 * Triggers 'moveFirst' action with 'Home' on 'root'.
 * Triggers 'moveLast' action with 'End' on 'root'.
 */
exports.accordionBehavior = function () { return ({
    attributes: {
        root: {
            role: 'presentation',
        },
    },
    keyActions: {
        root: {
            moveNext: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.ArrowDown }],
            },
            movePrevious: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.ArrowUp }],
            },
            moveFirst: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Home }],
            },
            moveLast: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.End }],
            },
        },
    },
}); };
