// @fluentui/accessibility/dist/commonjs/behaviors/HierarchicalTree/hierarchicalTreeItemBehavior.js
"use strict";

exports.__esModule = true;
exports.hierarchicalTreeItemBehavior = void 0;

var _keyboardKey = require("@fluentui/keyboard-key");

var _attributes = require("../../attributes");

/**
 * @description
 * Adds role 'treeitem' to a non-leaf item and 'none' to a leaf item.
 * Adds 'aria-expanded' with a value based on the 'open' prop if item is not a leaf.
 * Adds 'tabIndex' as '-1' if the item is not a leaf.
 *
 * @specification
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 * Triggers 'receiveFocus' action with 'ArrowLeft' on 'root', when has an opened subtree.
 * Triggers 'collapse' action with 'ArrowLeft' on 'root', when has an opened subtree.
 * Triggers 'expand' action with 'ArrowRight' on 'root', when has a closed subtree.
 * Triggers 'focusSubtree' action with 'ArrowRight' on 'root', when has an opened subtree.
 */
var hierarchicalTreeItemBehavior = function hierarchicalTreeItemBehavior(props) {
  var _ref;

  return {
    attributes: {
      root: Object.assign({
        role: 'none'
      }, props.hasItems && (_ref = {
        'aria-expanded': props.open,
        tabIndex: -1
      }, _ref[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _ref.role = 'treeitem', _ref))
    },
    keyActions: {
      root: Object.assign({
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      }, isSubtreeOpen(props) && {
        receiveFocus: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowLeft
          }]
        },
        collapse: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowLeft
          }]
        },
        focusSubtree: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowRight
          }]
        }
      }, !isSubtreeOpen(props) && {
        expand: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowRight
          }]
        }
      })
    }
  };
};

exports.hierarchicalTreeItemBehavior = hierarchicalTreeItemBehavior;

/** Checks if current tree item has a subtree and it is opened */
var isSubtreeOpen = function isSubtreeOpen(props) {
  var hasItems = props.hasItems,
      open = props.open;
  return !!(hasItems && open);
};
