// @fluentui/accessibility/dist/commonjs/behaviors/HierarchicalTree/hierarchicalTreeTitleBehavior.js
"use strict";

exports.__esModule = true;
exports.hierarchicalTreeTitleBehavior = void 0;

var _keyboardKey = require("@fluentui/keyboard-key");

var _attributes = require("../../attributes");

/**
 * @description
 * Adds role 'treeitem' if the title is a leaf node inside the tree.
 * Adds 'tabIndex' as '-1' if the title is a leaf node inside the tree.
 *
 * @specification
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 */
var hierarchicalTreeTitleBehavior = function hierarchicalTreeTitleBehavior(props) {
  var _ref;

  return {
    attributes: {
      root: Object.assign({}, !props.hasSubtree && (_ref = {
        tabIndex: -1
      }, _ref[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _ref.role = 'treeitem', _ref))
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      }
    }
  };
};

exports.hierarchicalTreeTitleBehavior = hierarchicalTreeTitleBehavior;
