// @fluentui/accessibility/dist/commonjs/behaviors/HierarchicalTree/hierarchicalTreeBehavior.js
"use strict";

exports.__esModule = true;
exports.hierarchicalTreeBehavior = void 0;

var _types = require("../../focusZone/types");

var _hierarchicalSubtreeBehavior = require("./hierarchicalSubtreeBehavior");

/**
 * @specification
 * Adds role 'tree' to 'root' slot.
 * Adds attribute 'aria-labelledby' based on the property 'aria-labelledby' to 'root' slot.
 * Provides arrow key navigation in vertical direction.
 * Triggers 'expandSiblings' action with '*' on 'root'.
 */
var hierarchicalTreeBehavior = function hierarchicalTreeBehavior(props) {
  var subtreeBehaviorData = (0, _hierarchicalSubtreeBehavior.hierarchicalSubtreeBehavior)({});
  return {
    attributes: {
      root: Object.assign({}, subtreeBehaviorData.attributes.root, {
        role: 'tree',
        'aria-labelledby': props['aria-labelledby']
      })
    },
    keyActions: {
      root: Object.assign({}, subtreeBehaviorData.keyActions.root)
    },
    focusZone: {
      props: {
        direction: _types.FocusZoneDirection.vertical
      }
    }
  };
};

exports.hierarchicalTreeBehavior = hierarchicalTreeBehavior;
