// @fluentui/react-northstar/dist/commonjs/themes/teams/components/HierarchicalTree/hierarchicalTreeItemStyles.js
"use strict";

exports.__esModule = true;
exports.hierarchicalTreeItemStyles = void 0;

var _utils = require("../../../../utils");

var _getBorderFocusStyles = require("../../getBorderFocusStyles");

var _HierarchicalTreeTitle = require("../../../../components/HierarchicalTree/HierarchicalTreeTitle");

var hierarchicalTreeItemStyles = {
  root: function root(_ref) {
    var _focusVisible;

    var siteVariables = _ref.theme.siteVariables;
    return Object.assign({
      listStyleType: 'none',
      padding: "0 0 0 " + (0, _utils.pxToRem)(1)
    }, (0, _getBorderFocusStyles.getBorderFocusStyles)({
      variables: siteVariables
    })[':focus'], {
      ':focus-visible': (_focusVisible = {}, _focusVisible["> ." + _HierarchicalTreeTitle.hierarchicalTreeTitleClassName] = Object.assign({
        position: 'relative'
      }, (0, _getBorderFocusStyles.getBorderFocusStyles)({
        variables: siteVariables
      })[':focus-visible']), _focusVisible)
    });
  }
};
exports.hierarchicalTreeItemStyles = hierarchicalTreeItemStyles;
