// @pnp/spfx-property-controls/lib/propertyFields/filePicker/placeHolderControl/PlaceholderComponent.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import * as React from 'react';
import { PrimaryButton } from 'office-ui-fabric-react/lib/Button';
import styles from './PlaceholderComponent.module.scss';
import { Icon } from 'office-ui-fabric-react/lib/components/Icon';
/**
 * Placeholder component
 */
var Placeholder = /** @class */ (function (_super) {
    __extends(Placeholder, _super);
    /**
     * Constructor
     */
    function Placeholder(props) {
        var _this = _super.call(this, props) || this;
        _this._crntElm = null;
        /**
         * Execute the onConfigure function
         */
        _this._handleBtnClick = function (event) {
            _this.props.onConfigure();
        };
        /**
         * Set the current zone width
         */
        _this._setZoneWidth = function () {
            _this.setState({
                width: _this._crntElm.clientWidth
            });
        };
        /**
         * Stores the current element
         */
        _this._linkElm = function (e) {
            _this._crntElm = e;
        };
        _this.state = {
            width: null
        };
        return _this;
    }
    /**
     * componentDidMount lifecycle hook
     */
    Placeholder.prototype.componentDidMount = function () {
        this._setZoneWidth();
    };
    /**
     * componentDidUpdate lifecycle hook
     * @param prevProps
     * @param prevState
     */
    Placeholder.prototype.componentDidUpdate = function (prevProps, prevState) {
        this._setZoneWidth();
    };
    /**
     * shouldComponentUpdate lifecycle hook
     * @param nextProps
     * @param nextState
     */
    Placeholder.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        return this.state.width !== nextState.width || this.props.hideButton !== nextProps.hideButton;
    };
    /**
     * Default React component render method
     */
    Placeholder.prototype.render = function () {
        return (React.createElement("div", { className: styles.placeholder + " " + (this.props.contentClassName ? this.props.contentClassName : ''), ref: this._linkElm },
            React.createElement("div", { className: styles.placeholderContainer },
                React.createElement("div", { className: styles.placeholderHead },
                    React.createElement("div", { className: styles.placeholderHeadContainer },
                        this.props.iconName && React.createElement(Icon, { iconName: this.props.iconName, className: styles.placeholderIcon }),
                        React.createElement("span", { className: styles.placeholderText + " " + ((this.state.width && this.state.width <= 380) ? styles.hide : "") }, this.props.iconText))),
                React.createElement("div", { className: styles.placeholderDescription },
                    React.createElement("span", { className: styles.placeholderDescriptionText }, this.props.description)),
                this.props.children,
                React.createElement("div", { className: styles.placeholderDescription }, (this.props.buttonLabel && !this.props.hideButton) &&
                    React.createElement(PrimaryButton, { text: this.props.buttonLabel, ariaLabel: this.props.buttonLabel, ariaDescription: this.props.description, onClick: this._handleBtnClick })))));
    };
    return Placeholder;
}(React.Component));
export { Placeholder };
