// @pnp/spfx-controls-react/lib/controls/fields/fieldLookupRenderer/FieldLookupRenderer.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { override } from '@microsoft/decorators';
import * as React from 'react';
import { css } from 'office-ui-fabric-react/lib/Utilities';
import { Dialog, DialogType } from 'office-ui-fabric-react/lib/Dialog';
import { Link } from 'office-ui-fabric-react/lib/Link';
import { Spinner, SpinnerSize } from 'office-ui-fabric-react/lib/Spinner';
import * as telemetry from '../../../common/telemetry';
import styles from './FieldLookupRenderer.module.scss';
import { IFrameDialog } from '../../iFrameDialog/IFrameDialog';
import { SPHelper } from '../../../Utilities';
/**
 * Field Lookup Renderer.
 * Used for:
 *   - Lookup, LookupMulti
 */
var FieldLookupRenderer = /** @class */ (function (_super) {
    __extends(FieldLookupRenderer, _super);
    function FieldLookupRenderer(props, state) {
        var _this = _super.call(this, props, state) || this;
        telemetry.track('FieldLookupRenderer', {});
        _this.state = {
            hideDialog: true,
            dispFormUrl: props.dispFormUrl
        };
        return _this;
    }
    FieldLookupRenderer.prototype.render = function () {
        var _this = this;
        var lookupLinks = this.props.lookups.map(function (lookup) {
            return React.createElement(Link, { onClick: _this._onClick.bind(_this, lookup), className: styles.lookup, style: _this.props.cssProps }, lookup.lookupValue);
        });
        return (React.createElement("div", { style: this.props.cssProps, className: css(this.props.className) },
            lookupLinks,
            !this.state.hideDialog && this.state.dispFormUrl && React.createElement(IFrameDialog, { url: this.state.lookupDispFormUrl, iframeOnLoad: this._onIframeLoaded.bind(this), hidden: this.state.hideDialog, onDismiss: this._onDialogDismiss.bind(this), modalProps: {
                    isBlocking: true,
                    containerClassName: styles.dialogContainer
                }, dialogContentProps: {
                    type: DialogType.close,
                    showCloseButton: true
                }, width: '570px', height: '250px' }),
            !this.state.hideDialog && !this.state.dispFormUrl && React.createElement(Dialog, { onDismiss: this._onDialogDismiss.bind(this), modalProps: {
                    isBlocking: true,
                    containerClassName: styles.dialogContainer
                }, dialogContentProps: {
                    type: DialogType.close,
                    showCloseButton: true
                } },
                React.createElement(Spinner, { size: SpinnerSize.large }))));
    };
    FieldLookupRenderer.prototype._onClick = function (lookup) {
        var _this = this;
        if (this.props.onClick) {
            var args = {
                lookup: lookup
            };
            this.props.onClick(args);
            return;
        }
        //
        // showing Display Form in the dialog
        //
        var dispFormUrl = this.state.dispFormUrl;
        if (dispFormUrl) {
            this.setState({
                lookupDispFormUrl: "" + dispFormUrl + (dispFormUrl.indexOf('?') === -1 ? '?' : '&') + "ID=" + lookup.lookupId + "&RootFolder=*&IsDlg=1",
                hideDialog: false
            });
        }
        else if (this.props.fieldId) {
            this.setState({
                hideDialog: false
            });
            SPHelper.getLookupFieldListDispFormUrl(this.props.fieldId, this.props.context).then(function (dispFormUrlValue) {
                var listDispFormUrl = dispFormUrlValue.toString();
                _this.setState(function (prevState, props) {
                    if (prevState.hideDialog) {
                        return;
                    }
                    return {
                        dispFormUrl: listDispFormUrl,
                        lookupDispFormUrl: listDispFormUrl + "&ID=" + lookup.lookupId + "&RootFolder=*&IsDlg=1"
                    };
                });
            });
        }
    };
    FieldLookupRenderer.prototype._onIframeLoaded = function (iframe) {
        //
        // some additional configuration to beutify content of the iframe
        //
        var iframeWindow = iframe.contentWindow;
        var iframeDocument = iframeWindow.document;
        var s4Workspace = iframeDocument.getElementById('s4-workspace');
        s4Workspace.style.height = iframe.style.height;
        s4Workspace.scrollIntoView();
    };
    FieldLookupRenderer.prototype._onDialogDismiss = function () {
        this.setState({
            hideDialog: true
        });
    };
    __decorate([
        override
    ], FieldLookupRenderer.prototype, "render", null);
    return FieldLookupRenderer;
}(React.Component));
export { FieldLookupRenderer };
