// @fluentui/accessibility/dist/dts/src/behaviors/HierarchicalTree/hierarchicalTreeBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var types_1 = require("../../focusZone/types");
var hierarchicalSubtreeBehavior_1 = require("./hierarchicalSubtreeBehavior");
/**
 * @specification
 * Adds role 'tree' to 'root' slot.
 * Adds attribute 'aria-labelledby' based on the property 'aria-labelledby' to 'root' slot.
 * Provides arrow key navigation in vertical direction.
 * Triggers 'expandSiblings' action with '*' on 'root'.
 */
exports.hierarchicalTreeBehavior = function (props) {
    var subtreeBehaviorData = hierarchicalSubtreeBehavior_1.hierarchicalSubtreeBehavior({});
    return {
        attributes: {
            root: tslib_1.__assign(tslib_1.__assign({}, subtreeBehaviorData.attributes.root), { role: 'tree', 'aria-labelledby': props['aria-labelledby'] }),
        },
        keyActions: {
            root: tslib_1.__assign({}, subtreeBehaviorData.keyActions.root),
        },
        focusZone: {
            props: {
                direction: types_1.FocusZoneDirection.vertical,
            },
        },
    };
};
