// @fluentui/react-northstar/dist/commonjs/components/HierarchicalTree/HierarchicalTreeTitle.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.HierarchicalTreeTitle = exports.hierarchicalTreeTitleClassName = void 0;

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@fluentui/accessibility");

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _utils = require("../../utils");

var _reactBindings = require("@fluentui/react-bindings");

var hierarchicalTreeTitleClassName = 'ui-hierarchicaltree__title';
exports.hierarchicalTreeTitleClassName = hierarchicalTreeTitleClassName;

/**
 * A TreeTitle renders a title of TreeItem.
 */
var HierarchicalTreeTitle = function HierarchicalTreeTitle(props) {
  var context = (0, _reactBindings.useFluentContext)();

  var _useTelemetry = (0, _reactBindings.useTelemetry)(HierarchicalTreeTitle.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var children = props.children,
      content = props.content,
      open = props.open,
      hasSubtree = props.hasSubtree,
      className = props.className,
      design = props.design,
      styles = props.styles,
      variables = props.variables;
  var ElementType = (0, _reactBindings.getElementType)(props);
  var unhandledProps = (0, _reactBindings.useUnhandledProps)(HierarchicalTreeTitle.handledProps, props);
  var getA11yProps = (0, _reactBindings.useAccessibility)(props.accessibility, {
    debugName: HierarchicalTreeTitle.displayName,
    actionHandlers: {
      performClick: function performClick(e) {
        e.preventDefault();
        handleClick(e);
      }
    },
    mapPropsToBehavior: function mapPropsToBehavior() {
      return {
        open: open,
        hasSubtree: hasSubtree
      };
    },
    rtl: context.rtl
  });

  var _useStyles = (0, _reactBindings.useStyles)(HierarchicalTreeTitle.displayName, {
    className: hierarchicalTreeTitleClassName,
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes;

  var handleClick = function handleClick(e) {
    (0, _invoke2.default)(props, 'onClick', e, props);
  };

  var element = /*#__PURE__*/React.createElement(ElementType, getA11yProps('root', Object.assign({
    className: classes.root,
    onClick: handleClick
  }, unhandledProps, _utils.rtlTextContainer.getAttributes({
    forElements: [children, content]
  }))), (0, _utils.childrenExist)(children) ? children : content);
  setEnd();
  return element;
};

exports.HierarchicalTreeTitle = HierarchicalTreeTitle;
HierarchicalTreeTitle.displayName = 'HierarchicalTreeTitle';
HierarchicalTreeTitle.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon(), {
  onClick: PropTypes.func,
  open: PropTypes.bool,
  hasSubtree: PropTypes.bool
});
HierarchicalTreeTitle.defaultProps = {
  as: 'a',
  accessibility: _accessibility.hierarchicalTreeTitleBehavior
};
HierarchicalTreeTitle.handledProps = Object.keys(HierarchicalTreeTitle.propTypes);
HierarchicalTreeTitle.create = (0, _utils.createShorthandFactory)({
  Component: HierarchicalTreeTitle,
  mappedProp: 'content'
});
