// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Menu/menuItemIndicatorStyles.js
"use strict";

exports.__esModule = true;
exports.menuItemIndicatorStyles = void 0;

var _utils = require("../../../../utils");

var _submenuIndicatorUrl = require("./submenuIndicatorUrl");

var menuItemIndicatorStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        rtl = _ref.rtl;
    return Object.assign({
      position: 'relative',
      float: 'right',
      left: (0, _utils.pxToRem)(12),
      userSelect: 'none',
      marginRight: (0, _utils.pxToRem)(4)
    }, p.inSubmenu && {
      position: 'absolute',
      top: (0, _utils.pxToRem)(6),
      right: (0, _utils.pxToRem)(2),
      left: 'unset'
    }, rtl && {
      transform: "scaleX(-1)"
    }, {
      content: '" "',
      display: 'block',
      overflow: 'hidden',
      height: (0, _utils.pxToRem)(16),
      width: (0, _utils.pxToRem)(16),
      backgroundSize: (0, _utils.pxToRem)(16),
      backgroundImage: (0, _submenuIndicatorUrl.submenuIndicatorUrl)(v.indicatorColor, p.vertical)
    }, p.active && Object.assign({
      backgroundImage: (0, _submenuIndicatorUrl.submenuIndicatorUrl)(v.activeIndicatorColor, p.vertical)
    }, p.primary && Object.assign({
      backgroundImage: (0, _submenuIndicatorUrl.submenuIndicatorUrl)(v.activePrimaryIndicatorColor, p.vertical)
    }, p.vertical && {
      backgroundImage: (0, _submenuIndicatorUrl.submenuIndicatorUrl)(v.activePrimaryVerticalIndicatorColor, p.vertical)
    })), p.underlined && {
      backgroundImage: (0, _submenuIndicatorUrl.submenuIndicatorUrl)(v.indicatorColor, p.vertical)
    }, p.iconOnly && {
      backgroundImage: (0, _submenuIndicatorUrl.submenuIndicatorUrl)(v.indicatorColor, p.vertical)
    });
  }
};
exports.menuItemIndicatorStyles = menuItemIndicatorStyles;
