// @pnp/spfx-controls-react/lib/controls/listItemAttachments/ListItemAttachments.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
// Joao Mendes November 2018, SPFx reusable Control ListItemAttachments
import * as React from 'react';
import { Dialog, DialogType, DialogFooter } from 'office-ui-fabric-react/lib/Dialog';
import { PrimaryButton, DefaultButton } from 'office-ui-fabric-react/lib/Button';
import { Label } from "office-ui-fabric-react/lib/Label";
import * as strings from 'ControlStrings';
import styles from './ListItemAttachments.module.scss';
import { UploadAttachment } from './UploadAttachment';
import { DocumentCard, DocumentCardActions, DocumentCardPreview } from 'office-ui-fabric-react/lib/DocumentCard';
import { ImageFit } from 'office-ui-fabric-react/lib/Image';
import SPservice from "../../services/SPService";
import { TooltipHost, DirectionalHint } from 'office-ui-fabric-react/lib/Tooltip';
import { Spinner, SpinnerSize } from 'office-ui-fabric-react/lib/Spinner';
import utilities from './utilities';
import { Placeholder } from "../placeholder";
import * as telemetry from '../../common/telemetry';
var ListItemAttachments = /** @class */ (function (_super) {
    __extends(ListItemAttachments, _super);
    function ListItemAttachments(props) {
        var _this = _super.call(this, props) || this;
        /**
         * Close the dialog
         */
        _this._closeDialog = function () {
            _this.setState({
                fireUpload: false,
                hideDialog: true,
                dialogMessage: '',
                file: null,
                deleteAttachment: false,
            });
            _this.loadAttachments();
        };
        /**
         * Attachment uploaded event handler
         */
        _this._onAttachmentUpload = function () {
            // load Attachments
            _this.loadAttachments();
        };
        /**
         * On delete attachment event handler
         *
         * @param file
         */
        _this.onDeleteAttachment = function (file) {
            _this.setState({
                fireUpload: false,
                hideDialog: false,
                deleteAttachment: true,
                file: file,
                dialogMessage: strings.ListItemAttachmentsconfirmDelete.replace('{0}', file.FileName),
            });
        };
        /**
         * Delete the attachment once it was confirmed
         */
        _this.onConfirmedDeleteAttachment = function () { return __awaiter(_this, void 0, void 0, function () {
            var file, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        file = this.state.file;
                        this.setState({
                            fireUpload: false,
                            disableButton: true,
                        });
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._spservice.deleteAttachment(file.FileName, this.props.listId, this.props.itemId, this.props.webUrl)];
                    case 2:
                        _a.sent();
                        this.setState({
                            fireUpload: false,
                            hideDialog: false,
                            deleteAttachment: false,
                            disableButton: false,
                            file: null,
                            dialogMessage: strings.ListItemAttachmentsfileDeletedMsg.replace('{0}', file.FileName),
                        });
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _a.sent();
                        this.setState({
                            fireUpload: false,
                            hideDialog: false,
                            file: null,
                            deleteAttachment: false,
                            dialogMessage: strings.ListItemAttachmentsfileDeleteError.replace('{0}', file.FileName).replace('{1}', error_1.message)
                        });
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        }); };
        telemetry.track('ReactListItemAttachments', {});
        _this.state = {
            file: null,
            hideDialog: true,
            dialogMessage: '',
            attachments: [],
            deleteAttachment: false,
            disableButton: false,
            showPlaceHolder: false,
            fireUpload: false
        };
        // Get SPService Factory
        _this._spservice = new SPservice(_this.props.context);
        _this._utilities = new utilities();
        return _this;
    }
    /**
     * componentDidMount lifecycle hook
     */
    ListItemAttachments.prototype.componentDidMount = function () {
        this.loadAttachments();
    };
    ListItemAttachments.prototype.loadAttachmentPreview = function (file) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this._utilities.GetFileImageUrl(file).then(function (previewImageUrl) {
                        return {
                            name: file.FileName,
                            previewImageSrc: previewImageUrl,
                            iconSrc: '',
                            imageFit: ImageFit.center,
                            width: 187,
                            height: 130,
                        };
                    })];
            });
        });
    };
    /**
     * Load Item Attachments
     */
    ListItemAttachments.prototype.loadAttachments = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this._spservice.getListItemAttachments(this.props.listId, this.props.itemId).then(function (files) {
                    var filePreviewImages = files.map(function (file) { return _this.loadAttachmentPreview(file); });
                    return Promise.all(filePreviewImages).then(function (filePreviews) {
                        _this.previewImages = {};
                        filePreviews.forEach(function (preview) {
                            _this.previewImages[preview.name] = preview;
                        });
                        _this.setState({
                            fireUpload: false,
                            hideDialog: true,
                            dialogMessage: '',
                            attachments: files,
                            showPlaceHolder: files.length === 0 ? true : false
                        });
                    });
                }).catch(function (error) {
                    _this.setState({
                        fireUpload: false,
                        hideDialog: false,
                        dialogMessage: strings.ListItemAttachmentserrorLoadAttachments.replace('{0}', error.message)
                    });
                });
                return [2 /*return*/];
            });
        });
    };
    /**
     * Default React render method
     */
    ListItemAttachments.prototype.render = function () {
        var _this = this;
        var openAttachmentsInNewWindow = this.props.openAttachmentsInNewWindow;
        return (React.createElement("div", { className: styles.ListItemAttachments },
            React.createElement(UploadAttachment, { listId: this.props.listId, itemId: this.props.itemId, disabled: this.props.disabled, context: this.props.context, onAttachmentUpload: this._onAttachmentUpload, fireUpload: this.state.fireUpload }),
            this.state.showPlaceHolder ?
                React.createElement(Placeholder, { iconName: 'Upload', iconText: strings.ListItemAttachmentslPlaceHolderIconText, description: strings.ListItemAttachmentslPlaceHolderDescription, buttonLabel: strings.ListItemAttachmentslPlaceHolderButtonLabel, hideButton: this.props.disabled, onConfigure: function () { return _this.setState({ fireUpload: true }); } })
                :
                    this.state.attachments.map(function (file) {
                        var fileName = file.FileName;
                        var previewImage = _this.previewImages[fileName];
                        return (React.createElement("div", { key: fileName, className: styles.documentCardWrapper },
                            React.createElement(TooltipHost, { content: fileName, calloutProps: { gapSpace: 0, isBeakVisible: true }, closeDelay: 200, directionalHint: DirectionalHint.rightCenter },
                                React.createElement(DocumentCard, { onClickHref: !openAttachmentsInNewWindow && file.ServerRelativeUrl + "?web=1", onClick: openAttachmentsInNewWindow && (function () { return window.open(file.ServerRelativeUrl + "?web=1", "_blank"); }), className: styles.documentCard },
                                    React.createElement(DocumentCardPreview, { previewImages: [previewImage] }),
                                    React.createElement(Label, { className: styles.fileLabel }, fileName),
                                    React.createElement(DocumentCardActions, { actions: [
                                            {
                                                iconProps: {
                                                    iconName: 'Delete',
                                                    title: strings.ListItemAttachmentsActionDeleteIconTitle,
                                                },
                                                title: strings.ListItemAttachmentsactionDeleteTitle,
                                                text: strings.ListItemAttachmentsactionDeleteTitle,
                                                disabled: _this.props.disabled,
                                                onClick: function (ev) {
                                                    ev.preventDefault();
                                                    ev.stopPropagation();
                                                    _this.onDeleteAttachment(file);
                                                }
                                            },
                                        ] })))));
                    }),
            React.createElement(Dialog, { hidden: this.state.hideDialog, type: DialogType.normal, onDismiss: this._closeDialog, dialogContentProps: {
                    type: DialogType.normal,
                    title: strings.ListItemAttachmentsdialogTitle,
                    subText: this.state.dialogMessage
                }, modalProps: {
                    isBlocking: true,
                    containerClassName: 'ms-dialogMainOverride'
                } },
                React.createElement(DialogFooter, null,
                    React.createElement("div", { style: { marginBottom: 7 } }, this.state.disableButton ? React.createElement(Spinner, { size: SpinnerSize.medium }) : null),
                    this.state.deleteAttachment ? (React.createElement(PrimaryButton, { disabled: this.state.disableButton, onClick: this.onConfirmedDeleteAttachment }, strings.ListItemAttachmentsdialogOKbuttonLabelOnDelete)) : null,
                    this.state.deleteAttachment ? (React.createElement(DefaultButton, { disabled: this.state.disableButton, onClick: this._closeDialog }, strings.ListItemAttachmentsdialogCancelButtonLabel))
                        : React.createElement(PrimaryButton, { onClick: this._closeDialog }, strings.ListItemAttachmentsdialogOKbuttonLabel)))));
    };
    return ListItemAttachments;
}(React.Component));
export { ListItemAttachments };
