// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Label/labelVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../../../utils");
var colorUtils_1 = require("../../../colorUtils");
exports.labelColorAreas = utils_1.stringLiteralsArray('foreground', 'background');
exports.labelVariables = function (siteVars) {
    var colorScheme = colorUtils_1.extendColorScheme(siteVars.colorScheme, {
        default: {
            background: 'rgba(0, 0, 0, 0.6)',
            foreground: 'rgb(232, 232, 232)',
        },
        brand: {
            background: siteVars.colorScheme.brand.foreground4,
        },
        red: {
            background: siteVars.colorScheme.red.foreground1,
        },
    });
    return {
        colorScheme: colorUtils_1.pickValuesFromColorScheme(colorScheme, exports.labelColorAreas),
        circularRadius: utils_1.pxToRem(9999),
        iconSize: utils_1.pxToRem(16),
        padding: "0 " + utils_1.pxToRem(4) + " 0 " + utils_1.pxToRem(4),
        startPaddingLeft: '0px',
        endPaddingRight: '0px',
        height: utils_1.pxToRem(20),
    };
};
