// @fluentui/react-northstar/dist/dts/src/components/Tree/TreeItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var _ = require("lodash");
var PropTypes = require("prop-types");
var React = require("react");
var react_component_ref_1 = require("@fluentui/react-component-ref");
var utils_1 = require("../../utils");
var TreeTitle_1 = require("./TreeTitle");
var utils_2 = require("./utils");
exports.treeItemClassName = 'ui-tree__item';
/**
 * A TreeItem renders an item of a Tree.
 *
 * @accessibility
 * Implements [ARIA TreeView](https://www.w3.org/TR/wai-aria-practices-1.1/#TreeView) design pattern.
 */
exports.TreeItem = function (props) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.TreeItem.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, children = props.children, className = props.className, contentRef = props.contentRef, design = props.design, title = props.title, renderItemTitle = props.renderItemTitle, expanded = props.expanded, level = props.level, index = props.index, styles = props.styles, variables = props.variables, treeSize = props.treeSize, selectionIndicator = props.selectionIndicator, selectableParent = props.selectableParent, selected = props.selected, selectable = props.selectable, indeterminate = props.indeterminate, id = props.id;
    var hasSubtreeItem = utils_2.hasSubtree(props);
    var _b = React.useContext(utils_2.TreeContext), onFocusParent = _b.onFocusParent, onSiblingsExpand = _b.onSiblingsExpand, onFocusFirstChild = _b.onFocusFirstChild, onTitleClick = _b.onTitleClick;
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        actionHandlers: {
            performClick: function (e) {
                e.preventDefault();
                e.stopPropagation();
                handleTitleClick(e);
            },
            focusParent: function (e) {
                e.preventDefault();
                e.stopPropagation();
                handleFocusParent(e);
            },
            collapse: function (e) {
                e.preventDefault();
                e.stopPropagation();
                handleTitleClick(e);
            },
            expand: function (e) {
                e.preventDefault();
                e.stopPropagation();
                handleTitleClick(e);
            },
            focusFirstChild: function (e) {
                e.preventDefault();
                e.stopPropagation();
                handleFocusFirstChild(e);
            },
            expandSiblings: function (e) {
                e.preventDefault();
                e.stopPropagation();
                handleSiblingsExpand(e);
            },
            performSelection: function (e) {
                e.preventDefault();
                e.stopPropagation();
                handleSelection(e);
            },
        },
        debugName: exports.TreeItem.displayName,
        mapPropsToBehavior: function () { return ({
            expanded: expanded,
            level: level,
            index: index,
            hasSubtree: hasSubtreeItem,
            treeSize: treeSize,
            selected: selected,
            selectable: selectable,
            selectableParent: selectableParent,
        }); },
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.TreeItem.displayName, {
        className: exports.treeItemClassName,
        mapPropsToStyles: function () { return ({
            level: level,
        }); },
        mapPropsToInlineStyles: function () { return ({ className: className, design: design, styles: styles, variables: variables }); },
        rtl: context.rtl,
    }).classes;
    var handleSelection = function (e) {
        onTitleClick(e, props, true);
        _.invoke(props, 'onTitleClick', e, props);
    };
    var handleTitleClick = function (e) {
        onTitleClick(e, props);
        _.invoke(props, 'onTitleClick', e, props);
    };
    var handleFocusFirstChild = function (e) {
        _.invoke(props, 'onFocusFirstChild', e, props);
        onFocusFirstChild(props.id);
    };
    var handleFocusParent = function (e) {
        _.invoke(props, 'onFocusParent', e, props);
        onFocusParent(props.parent);
    };
    var handleSiblingsExpand = function (e) {
        _.invoke(props, 'onSiblingsExpand', e, props);
        onSiblingsExpand(e, props);
    };
    var handleTitleOverrides = function (predefinedProps) { return ({
        onClick: function (e, titleProps) {
            var _a, _b, _c;
            if (selectable && ((_c = (_b = (_a = e) === null || _a === void 0 ? void 0 : _a.target) === null || _b === void 0 ? void 0 : _b.className) === null || _c === void 0 ? void 0 : _c.includes(TreeTitle_1.treeTitleSlotClassNames.indicator))) {
                handleSelection(e);
            }
            else {
                handleTitleClick(e);
            }
            _.invoke(predefinedProps, 'onClick', e, titleProps);
        },
    }); };
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.TreeItem.handledProps, props);
    var element = (React.createElement(ElementType, tslib_1.__assign({}, getA11Props('root', tslib_1.__assign(tslib_1.__assign({ className: classes.root, id: id,
        selected: selected }, utils_1.rtlTextContainer.getAttributes({ forElements: [children] })), unhandledProps))), utils_1.childrenExist(children)
        ? children
        : TreeTitle_1.TreeTitle.create(title, {
            defaultProps: function () {
                return getA11Props('title', tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({ hasSubtree: hasSubtreeItem, as: hasSubtreeItem ? 'span' : 'a', level: level,
                    treeSize: treeSize,
                    expanded: expanded,
                    index: index,
                    selected: selected,
                    selectable: selectable }, (hasSubtreeItem && !selectableParent && { selectable: false })), (selectableParent && { indeterminate: indeterminate })), { selectableParent: selectableParent,
                    selectionIndicator: selectionIndicator }));
            },
            render: renderItemTitle,
            overrideProps: handleTitleOverrides,
        })));
    var elementWithRef = contentRef ? React.createElement(react_component_ref_1.Ref, { innerRef: contentRef }, element) : element;
    setEnd();
    return elementWithRef;
};
exports.TreeItem.displayName = 'TreeItem';
exports.TreeItem.propTypes = tslib_1.__assign(tslib_1.__assign({}, utils_1.commonPropTypes.createCommon({
    content: false,
})), { contentRef: customPropTypes.ref, id: PropTypes.string.isRequired, index: PropTypes.number, items: customPropTypes.collectionShorthand, level: PropTypes.number, onFocusFirstChild: PropTypes.func, onFocusParent: PropTypes.func, onTitleClick: PropTypes.func, onSiblingsExpand: PropTypes.func, expanded: PropTypes.bool, parent: PropTypes.string, renderItemTitle: PropTypes.func, treeSize: PropTypes.number, title: customPropTypes.itemShorthand, selectionIndicator: customPropTypes.shorthandAllowingChildren, selected: PropTypes.bool, selectable: PropTypes.bool, selectableParent: PropTypes.bool, indeterminate: PropTypes.bool });
exports.TreeItem.defaultProps = {
    accessibility: accessibility_1.treeItemBehavior,
};
exports.TreeItem.handledProps = Object.keys(exports.TreeItem.propTypes);
exports.TreeItem.create = utils_1.createShorthandFactory({
    Component: exports.TreeItem,
    mappedProp: 'title',
});
