// @pnp/spfx-property-controls/lib/propertyFields/choiceGroupWithCallout/PropertyFieldChoiceGroupWithCallout.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { PropertyPaneFieldType } from '@microsoft/sp-webpart-base';
import omit from 'lodash/omit';
import PropertyFieldToggleWithCalloutHost from './PropertyFieldChoiceGroupWithCalloutHost';
var PropertyFieldChoiceGroupWithCalloutBuilder = /** @class */ (function () {
    function PropertyFieldChoiceGroupWithCalloutBuilder(_targetProperty, _properties) {
        this.type = PropertyPaneFieldType.Custom;
        this.targetProperty = _targetProperty;
        this.properties = _properties;
        this.properties.onRender = this._render.bind(this);
        this.properties.onDispose = this._dispose.bind(this);
    }
    PropertyFieldChoiceGroupWithCalloutBuilder.prototype._render = function (elem, context, changeCallback) {
        // IPropertyPaneChoiceGroupOption should be manually converted to IChoiceGroupOption
        var options = this.properties.options.map(function (o) {
            return __assign({}, omit(o, ['key', 'iconProps']), { iconProps: o.iconProps && {
                    iconName: o.iconProps.officeFabricIconFontName
                }, key: o.key.toString() });
        });
        var props = this.properties;
        var element = React.createElement(PropertyFieldToggleWithCalloutHost, __assign({}, omit(props, ['options']), { options: options, onChanged: this._onChanged.bind(this) }));
        ReactDOM.render(element, elem);
        if (changeCallback) {
            this._onChangeCallback = changeCallback;
        }
    };
    PropertyFieldChoiceGroupWithCalloutBuilder.prototype._dispose = function (elem) {
        ReactDOM.unmountComponentAtNode(elem);
    };
    PropertyFieldChoiceGroupWithCalloutBuilder.prototype._onChanged = function (option) {
        if (this._onChangeCallback) {
            this._onChangeCallback(this.targetProperty, option.key);
        }
    };
    return PropertyFieldChoiceGroupWithCalloutBuilder;
}());
export function PropertyFieldChoiceGroupWithCallout(targetProperty, properties) {
    return new PropertyFieldChoiceGroupWithCalloutBuilder(targetProperty, __assign({}, properties, { onRender: null, onDispose: null }));
}
