// @fluentui/react-northstar/dist/dts/src/themes/teams/components/TextArea/textAreaStyles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
exports.textAreaStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables;
        return (tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({ margin: v.margin, height: v.height, backgroundColor: v.backgroundColor }, (p.inverted && {
            backgroundColor: v.invertedBackgroundColor,
        })), { color: v.fontColor, borderColor: v.borderColor, borderRadius: v.borderRadius, borderStyle: 'solid', borderWidth: v.borderWidth, outline: 0, padding: v.padding, resize: p.resize || 'none' }), (p.fluid && {
            width: '100%',
        })), (p.disabled && {
            pointerEvents: 'none',
            backgroundColor: v.disabledBackgroundColor,
            color: v.disabledColor,
        })), { '::placeholder': {
                color: v.placeholderColor,
                opacity: 1,
            }, ':focus': {
                borderColor: v.borderColorFocus,
            } }));
    },
};
