// @fluentui/react-northstar/dist/dts/src/themes/teams/components/TextArea/textAreaVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../../../utils");
exports.textAreaVariables = function (siteVars) { return ({
    margin: '0',
    padding: utils_1.pxToRem(7) + " " + utils_1.pxToRem(12),
    borderColor: 'transparent',
    borderRadius: utils_1.pxToRem(3) + " " + utils_1.pxToRem(3) + " " + utils_1.pxToRem(2) + " " + utils_1.pxToRem(2),
    borderWidth: "0 0 " + utils_1.pxToRem(2) + " 0",
    backgroundColor: siteVars.colorScheme.default.background2,
    invertedBackgroundColor: siteVars.colorScheme.default.background,
    placeholderColor: siteVars.colorScheme.default.foreground1,
    disabledBackgroundColor: siteVars.colorScheme.default.backgroundColorDisabled,
    disabledColor: siteVars.colorScheme.default.colorDisabled,
    fontColor: siteVars.colorScheme.default.foreground,
    fontSize: siteVars.fontSizes.medium,
    borderColorFocus: "transparent transparent " + siteVars.colorScheme.brand.borderFocus1 + " transparent",
    height: 'auto',
}); };
