// @fluentui/react-northstar/dist/dts/test/utils/domEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// But at runtime it has to be imported like this
var simulant = require('simulant');
// ----------------------------------------
// Simulate DOM Events on real DOM nodes
// ----------------------------------------
/**
 * Generic method for dispatching an event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param eventType - A DOMString
 * @param data - Additional event data.
 * @returns The event
 */
exports.fire = function (node, eventType, data) {
    if (data === void 0) { data = {}; }
    var DOMNode = typeof node === 'string' ? document.querySelector(node) : node;
    var event = simulant(eventType, data);
    return simulant.fire(DOMNode, event);
};
/**
 * Dispatch a 'click' event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param data - Additional event data.
 * @returns The event
 */
exports.click = function (node, data) { return exports.fire(node, 'click', data); };
/**
 * Dispatch a 'keydown' event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param data - Additional event data.
 * @returns The event
 */
exports.keyDown = function (node, data) { return exports.fire(node, 'keydown', data); };
/**
 * Dispatch a 'mouseenter' event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param data - Additional event data.
 * @returns The event
 */
exports.mouseEnter = function (node, data) { return exports.fire(node, 'mouseenter', data); };
/**
 * Dispatch a 'mouseleave' event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param data - Additional event data.
 * @returns The event
 */
exports.mouseLeave = function (node, data) { return exports.fire(node, 'mouseleave', data); };
/**
 * Dispatch a 'mouseover' event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param data - Additional event data.
 * @returns The event
 */
exports.mouseOver = function (node, data) { return exports.fire(node, 'mouseover', data); };
/**
 * Dispatch a 'mouseup' event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param data - Additional event data.
 * @returns The event
 */
exports.mouseUp = function (node, data) { return exports.fire(node, 'mouseup', data); };
/**
 * Dispatch a 'resize' event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param data - Additional event data.
 * @returns The event
 */
exports.resize = function (node, data) { return exports.fire(node, 'resize', data); };
/**
 * Dispatch a 'scroll' event on a DOM node.
 * @param node - A querySelector string or DOM node.
 * @param data - Additional event data.
 * @returns The event
 */
exports.scroll = function (node, data) { return exports.fire(node, 'scroll', data); };
