// @pnp/spfx-controls-react/lib/services/SPService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { LibsOrderBy } from "./ISPService";
import { SPHttpClient } from "@microsoft/sp-http";
var SPService = /** @class */ (function () {
    function SPService(_context, webAbsoluteUrl) {
        this._context = _context;
        this._webAbsoluteUrl = webAbsoluteUrl ? webAbsoluteUrl : this._context.pageContext.web.absoluteUrl;
    }
    /**
     * Get lists or libraries
     *
     * @param options
     */
    SPService.prototype.getLibs = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var filtered, queryUrl, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        queryUrl = this._webAbsoluteUrl + "/_api/web/lists?$select=Title,id,BaseTemplate";
                        if (options.orderBy) {
                            queryUrl += "&$orderby=" + (options.orderBy === LibsOrderBy.Id ? 'Id' : 'Title');
                        }
                        if (options.filter) {
                            queryUrl += "&$filter=" + encodeURIComponent(options.filter);
                        }
                        else {
                            if (options.baseTemplate) {
                                queryUrl += "&$filter=BaseTemplate eq " + options.baseTemplate;
                                filtered = true;
                            }
                            if (options.includeHidden === false) {
                                queryUrl += filtered ? ' and Hidden eq false' : '&$filter=Hidden eq false';
                                filtered = true;
                            }
                        }
                        return [4 /*yield*/, this._context.spHttpClient.get(queryUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3: return [2 /*return*/, null];
                }
            });
        });
    };
    /**
     * Get List Items
     */
    SPService.prototype.getListItems = function (filterText, listId, internalColumnName, keyInternalColumnName, webUrl, filter, substringSearch) {
        if (substringSearch === void 0) { substringSearch = false; }
        return __awaiter(this, void 0, void 0, function () {
            var returnItems, filterStr, webAbsoluteUrl, apiUrl, data, results, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        filterStr = substringSearch ? // JJ - 20200613 - find by substring as an option
                            "substringof('" + encodeURIComponent(filterText.replace("'", "''")) + "'," + internalColumnName + ")" + (filter ? ' and ' + filter : '')
                            : "startswith(" + internalColumnName + ",'" + encodeURIComponent(filterText.replace("'", "''")) + "')" + (filter ? ' and ' + filter : '');
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 5, , 6]);
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items?$select=" + (keyInternalColumnName || 'Id') + "," + internalColumnName + "&$filter=" + filterStr;
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 2:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 4];
                        return [4 /*yield*/, data.json()];
                    case 3:
                        results = _a.sent();
                        if (results && results.value && results.value.length > 0) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 4;
                    case 4: return [2 /*return*/, []];
                    case 5:
                        error_1 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_1)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
  * Gets list items for list item picker
  * @param filterText
  * @param listId
  * @param internalColumnName
  * @param [keyInternalColumnName]
  * @param [webUrl]
  * @param [filterList]
  * @returns list items for list item picker
  */
    SPService.prototype.getListItemsForListItemPicker = function (filterText, listId, internalColumnName, keyInternalColumnName, webUrl, filterList) {
        return __awaiter(this, void 0, void 0, function () {
            var _filter, costumfilter, _top, webAbsoluteUrl, apiUrl, data, results, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _filter = "$filter=startswith(" + internalColumnName + ",'" + encodeURIComponent(filterText.replace("'", "''")) + "') ";
                        costumfilter = filterList
                            ? "and " + filterList
                            : "";
                        _top = " &$top=2000";
                        // test wild character "*"  if "*" load first 30 items
                        if ((filterText.trim().indexOf("*") == 0 &&
                            filterText.trim().length == 1) ||
                            filterText.trim().length == 0) {
                            _filter = "";
                            costumfilter = filterList ? "$filter=" + filterList + "&" : "";
                            _top = "$top=500";
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 5, , 6]);
                        webAbsoluteUrl = !webUrl
                            ? this._webAbsoluteUrl
                            : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items?$orderby=" + internalColumnName + "&$select=" + (keyInternalColumnName ||
                            "Id") + "," + internalColumnName + "&" + _filter + costumfilter + _top;
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 2:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 4];
                        return [4 /*yield*/, data.json()];
                    case 3:
                        results = _a.sent();
                        if (results &&
                            results.value &&
                            results.value.length > 0) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 4;
                    case 4: return [2 /*return*/, []];
                    case 5:
                        error_2 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_2)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get list item attachments
     *
     * @param listId
     * @param itemId
     * @param webUrl
     */
    SPService.prototype.getListItemAttachments = function (listId, itemId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(@itemId)/AttachmentFiles?@listId=guid'" + encodeURIComponent(listId) + "'&@itemId=" + encodeURIComponent(String(itemId));
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results && results.value) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_3 = _a.sent();
                        console.dir(error_3);
                        return [2 /*return*/, Promise.reject(error_3)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Delete attachment
     *
     * @param fileName
     * @param listId
     * @param itemId
     * @param webUrl
     */
    SPService.prototype.deleteAttachment = function (fileName, listId, itemId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var spOpts, webAbsoluteUrl, apiUrl, data, error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        spOpts = {
                            headers: { "X-HTTP-Method": 'DELETE', }
                        };
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(@itemId)/AttachmentFiles/getByFileName(@fileName)/RecycleObject?@listId=guid'" + encodeURIComponent(listId) + "'&@itemId=" + encodeURIComponent(String(itemId)) + "&@fileName='" + encodeURIComponent(fileName.replace(/'/g, "''")) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, SPHttpClient.configurations.v1, spOpts)];
                    case 1:
                        data = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        error_4 = _a.sent();
                        console.dir(error_4);
                        return [2 /*return*/, Promise.reject(error_4)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Add attachment
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param file
     * @param webUrl
     */
    SPService.prototype.addAttachment = function (listId, itemId, fileName, file, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var fileExists, spOpts, webAbsoluteUrl, apiUrl, data, error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        // Remove special characters in FileName
                        fileName = fileName.replace(/[^\.\w\s\&\-]/gi, '');
                        return [4 /*yield*/, this.checkAttachmentExists(listId, itemId, fileName, webUrl)];
                    case 1:
                        fileExists = _a.sent();
                        if (!fileExists) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.deleteAttachment(fileName, listId, itemId, webUrl)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3:
                        spOpts = {
                            body: file
                        };
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(@itemId)/AttachmentFiles/add(FileName=@fileName)?@listId=guid'" + encodeURIComponent(listId) + "'&@itemId=" + encodeURIComponent(String(itemId)) + "&@fileName='" + encodeURIComponent(fileName.replace(/'/g, "''")) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, SPHttpClient.configurations.v1, spOpts)];
                    case 4:
                        data = _a.sent();
                        return [2 /*return*/];
                    case 5:
                        error_5 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_5)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get attachement for list item
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param webUrl
     */
    SPService.prototype.getAttachment = function (listId, itemId, fileName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, file;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(@itemId)/AttachmentFiles/GetByFileBame(@fileName))?@listId=guid'" + encodeURIComponent(listId) + "'&@itemId=" + encodeURIComponent(String(itemId)) + "&@fileName='" + encodeURIComponent(fileName.replace(/'/g, "''")) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        file = _a.sent();
                        if (file) {
                            return [2 /*return*/, file];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                }
            });
        });
    };
    /**
     * Check if the attachment exists
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param webUrl
     */
    SPService.prototype.checkAttachmentExists = function (listId, itemId, fileName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var listServerRelativeUrl, webAbsoluteUrl, fileServerRelativeUrl, apiUrl, data, results, error_6;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        return [4 /*yield*/, this.getListServerRelativeUrl(listId, webUrl)];
                    case 1:
                        listServerRelativeUrl = _a.sent();
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        fileServerRelativeUrl = listServerRelativeUrl + "/Attachments/" + itemId + "/" + fileName;
                        apiUrl = webAbsoluteUrl + "/_api/web/getfilebyserverrelativeurl(@url)/Exists?@url='" + encodeURIComponent(fileServerRelativeUrl.replace(/'/g, "''")) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 2:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 4];
                        return [4 /*yield*/, data.json()];
                    case 3:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 4;
                    case 4: return [2 /*return*/, false];
                    case 5:
                        error_6 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_6)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get the list name
     *
     * @param listId
     * @param webUrl
     */
    SPService.prototype.getListName = function (listId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/RootFolder/Name?@listId=guid'" + encodeURIComponent(listId) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get the list server relative url
     *
     * @param listId
     * @param webUrl
     */
    SPService.prototype.getListServerRelativeUrl = function (listId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/RootFolder/ServerRelativeUrl?@listId=guid'" + encodeURIComponent(listId) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return SPService;
}());
export default SPService;
