// @pnp/spfx-property-controls/lib/propertyFields/codeEditor/CodeFormatter.js
import { CSSFormatter } from './CSSFormatter';
import { ScriptFormatter } from './ScriptFormatter';
/**
 * Helper class to format the code.
 */
var CodeFormatter = /** @class */ (function () {
    function CodeFormatter() {
    }
    /**
     * Formats CSS code
     * @param rawCSS
     */
    CodeFormatter.prototype.formatCSS = function (rawCSS) {
        var cssFormatter = new CSSFormatter();
        return cssFormatter.css_beautify(rawCSS, null);
    };
    /**
     * Formats JavaScript code
     * @param rawScript
     */
    CodeFormatter.prototype.formatScript = function (rawScript) {
        var scriptFormatter = new ScriptFormatter();
        return scriptFormatter.js_beautify(rawScript, null);
    };
    /**
     * Formats XML code
     * @param rawXML
     */
    CodeFormatter.prototype.formatXML = function (rawXML) {
        var tab = '\t';
        var result = '';
        var indent = '';
        rawXML.split(/>\s*</).forEach(function (element) {
            if (element.match(/^\/\w/)) {
                indent = indent.substring(tab.length);
            }
            result += indent + '<' + element + '>\r\n';
            if (element.match(/^<?\w[^>]*[^\/]$/)) {
                indent += tab;
            }
        });
        return result.substring(1, result.length - 3);
    };
    /**
     * Formats HTML code
     * @param rawHTML
     */
    CodeFormatter.prototype.formatHTML = function (rawHTML) {
        return this.formatXML(rawHTML);
    };
    /**
     * Formats JSON code
     * @param rawJSON
     */
    CodeFormatter.prototype.formatJSON = function (rawJSON) {
        return JSON.stringify(JSON.parse(rawJSON), null, 2);
    };
    return CodeFormatter;
}());
export { CodeFormatter };
