// @fluentui/react-northstar/dist/dts/src/components/Menu/MenuItemIndicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_bindings_1 = require("@fluentui/react-bindings");
var accessibility_1 = require("@fluentui/accessibility");
var PropTypes = require("prop-types");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.menuItemIndicatorClassName = 'ui-menu__itemindicator';
/**
 * A MenuItemIndicator allows a user to have a dedicated component that can be targeted from the theme.
 */
exports.MenuItemIndicator = react_bindings_1.compose(Box_1.Box, {
    className: exports.menuItemIndicatorClassName,
    displayName: 'MenuItemIndicator',
    mapPropsToStylesProps: function (props) { return ({
        iconOnly: props.iconOnly,
        vertical: props.vertical,
        inSubmenu: props.inSubmenu,
        active: props.active,
        primary: props.primary,
        underlined: props.underlined,
    }); },
    handledProps: ['iconOnly', 'vertical', 'inSubmenu', 'active', 'primary', 'underlined'],
    overrideStyles: true,
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.MenuItemIndicator.defaultProps = {
    as: 'span',
    accessibility: accessibility_1.indicatorBehavior,
};
exports.MenuItemIndicator.propTypes = tslib_1.__assign(tslib_1.__assign({}, utils_1.commonPropTypes.createCommon()), { iconOnly: PropTypes.bool, vertical: PropTypes.bool, inSubmenu: PropTypes.bool, active: PropTypes.bool, primary: PropTypes.bool, underlined: PropTypes.bool });
