// @fluentui/react-northstar/dist/dts/src/components/Menu/MenuItemContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_bindings_1 = require("@fluentui/react-bindings");
var PropTypes = require("prop-types");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.menuItemContentClassName = 'ui-menu__itemcontent';
/**
 * A MenuItemContent allows a user to have a dedicated component that can be targeted from the theme.
 */
exports.MenuItemContent = react_bindings_1.compose(Box_1.Box, {
    className: exports.menuItemContentClassName,
    displayName: 'MenuItemContent',
    mapPropsToStylesProps: function (props) { return ({
        hasMenu: props.hasMenu,
        hasIcon: props.hasIcon,
        vertical: props.vertical,
        inSubmenu: props.inSubmenu,
    }); },
    handledProps: ['hasMenu', 'hasIcon', 'vertical', 'inSubmenu'],
    overrideStyles: true,
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.MenuItemContent.defaultProps = {
    as: 'span',
};
exports.MenuItemContent.propTypes = tslib_1.__assign(tslib_1.__assign({}, utils_1.commonPropTypes.createCommon()), { hasIcon: PropTypes.bool, hasMenu: PropTypes.bool, vertical: PropTypes.bool, inSubmenu: PropTypes.bool });
