// @fluentui/react-northstar/dist/dts/src/components/Menu/MenuItemWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.menuItemWrapperClassName = 'ui-menu__itemwrapper';
/**
 * A MenuItemWrapper allows a user to have a dedicated component that can be targeted from the theme.
 */
exports.MenuItemWrapper = react_bindings_1.compose(Box_1.Box, {
    className: exports.menuItemWrapperClassName,
    displayName: 'MenuItemWrapper',
    mapPropsToStylesProps: function (props) { return ({
        active: props.active,
        disabled: props.disabled,
        iconOnly: props.iconOnly,
        isFromKeyboard: props.isFromKeyboard,
        pills: props.pills,
        pointing: props.pointing,
        secondary: props.secondary,
        underlined: props.underlined,
        vertical: props.vertical,
        primary: props.primary,
    }); },
    handledProps: [
        'active',
        'disabled',
        'iconOnly',
        'isFromKeyboard',
        'pills',
        'pointing',
        'secondary',
        'underlined',
        'vertical',
        'primary',
    ],
    overrideStyles: true,
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.MenuItemWrapper.defaultProps = {
    as: 'li',
};
exports.MenuItemWrapper.propTypes = tslib_1.__assign(tslib_1.__assign({}, utils_1.commonPropTypes.createCommon()), { active: PropTypes.bool, disabled: PropTypes.bool, iconOnly: PropTypes.bool, isFromKeyboard: PropTypes.bool, pills: PropTypes.bool, pointing: PropTypes.oneOf(['start', 'end', true, false]), primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]), secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]), underlined: PropTypes.bool, vertical: PropTypes.bool });
