// @fluentui/accessibility/dist/commonjs/behaviors/Tree/treeBehavior.js
"use strict";

exports.__esModule = true;
exports.treeBehavior = void 0;

var _keyboardKey = require("@fluentui/keyboard-key");

var _types = require("../../focusZone/types");

var _treeItemBehavior = require("./treeItemBehavior");

/**
 * @specification
 * Adds role 'tree' to 'root' slot.
 * Adds attribute 'tabIndex=-1' to 'root' slot.
 * Adds attribute 'aria-labelledby' based on the property 'aria-labelledby' to 'root' slot.
 * Provides arrow key navigation in vertical direction.
 * Triggers 'expandSiblings' action with '*' on 'root'.
 */
var treeBehavior = function treeBehavior(props) {
  return {
    attributes: {
      root: {
        role: 'tree',
        'aria-labelledby': props['aria-labelledby'],
        tabIndex: -1
      }
    },
    keyActions: {
      root: {
        expandSiblings: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey['*']
          }]
        }
      }
    },
    focusZone: {
      props: {
        direction: _types.FocusZoneDirection.vertical
      }
    },
    childBehaviors: {
      item: _treeItemBehavior.treeItemBehavior
    }
  };
};

exports.treeBehavior = treeBehavior;
