// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Alert/alertVariables.js
"use strict";

exports.__esModule = true;
exports.alertVariables = void 0;

var _utils = require("../../../../utils");

var alertVariables = function alertVariables(siteVars) {
  var minHeight = (0, _utils.pxToRem)(28);
  return {
    borderStyle: 'solid',
    borderWidth: '1px',
    borderRadius: (0, _utils.pxToRem)(3),
    backgroundColor: siteVars.colors.grey[50],
    // $app-white
    borderColor: siteVars.colors.grey[250],
    color: siteVars.colors.grey[500],
    fontWeight: siteVars.fontWeightRegular,
    minHeight: minHeight,
    padding: "0 " + (0, _utils.pxToRem)(16),
    actionsMargin: (0, _utils.pxToRem)(5),
    hoverBorderColor: undefined,
    hoverBackgroundColor: undefined,
    focusBackgroundColor: undefined,
    dismissActionHoverBorderRadius: undefined,
    dismissActionHoverBorderWidth: undefined,
    dismissActionHoverInnerBorderColor: undefined,
    dismissActionHoverOuterBorderColor: undefined,
    dismissActionHoverZIndex: undefined,
    dismissActionSize: minHeight,
    dismissActionColor: undefined,
    dismissiblePadding: "0 0 0 " + (0, _utils.pxToRem)(16),
    dangerColor: siteVars.colors.red[400],
    dangerBackgroundColor: siteVars.colors.red[50],
    dangerBorderColor: siteVars.colors.red[100],
    oof: false,
    oofColor: siteVars.colors.pink[600],
    oofBackgroundColor: siteVars.colors.pink[50],
    oofBorderColor: siteVars.colors.pink[100],
    infoColor: siteVars.colors.grey[500],
    infoBackgroundColor: siteVars.colors.grey[150],
    infoBorderColor: siteVars.colors.grey[200],
    urgent: false,
    urgentColor: siteVars.colors.white,
    urgentBackgroundColor: siteVars.colors.red[400],
    urgentBorderColor: siteVars.colors.red[400],
    headerFontWeight: siteVars.fontWeightBold,
    headerMargin: "0 " + (0, _utils.pxToRem)(10) + " 0 0",
    iconMargin: "0 " + (0, _utils.pxToRem)(10) + " 0 0",
    iconSize: (0, _utils.pxToRem)(16),
    dismissActionBackgroundColor: 'transparent',
    dismissActionBorderRadius: siteVars.borderRadius,
    dismissActionBorderColor: 'transparent',
    dismissActionColorHover: siteVars.colorScheme.brand.foregroundHover,
    dismissActionBackgroundColorHover: siteVars.colorScheme.default.backgroundHover2,
    dismissActionBorderColorHover: siteVars.colorScheme.default.borderHover,
    dismissActionContentFontWeight: siteVars.fontWeightSemibold,
    dismissActionBackgroundColorFocus: undefined,
    dismissActionBorderColorFocus: undefined,
    dismissActionColorFocus: undefined,
    dismissActionColorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    dismissActionBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
    dismissActionBorderColorDisabled: 'transparent',
    dismissActionIndicatorSize: (0, _utils.pxToRem)(16),
    focusBorderRadius: siteVars.borderRadius,
    focusBorderWidth: siteVars.borderWidth,
    focusInnerBorderColor: siteVars.focusInnerBorderColor,
    focusOuterBorderColor: siteVars.focusOuterBorderColor,
    focusBorderZIndex: siteVars.zIndexes.foreground
  };
};

exports.alertVariables = alertVariables;
