// @fluentui/react-northstar/dist/dts/test/specs/commonTests/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var implementsCollectionShorthandProp_1 = require("./implementsCollectionShorthandProp");
exports.implementsCollectionShorthandProp = implementsCollectionShorthandProp_1.implementsCollectionShorthandProp;
var implementsShorthandProp_1 = require("./implementsShorthandProp");
exports.implementsShorthandProp = implementsShorthandProp_1.implementsShorthandProp;
var implementsWrapperProp_1 = require("./implementsWrapperProp");
exports.implementsWrapperProp = implementsWrapperProp_1.implementsWrapperProp;
var handlesAccessibility_1 = require("./handlesAccessibility");
exports.handlesAccessibility = handlesAccessibility_1.handlesAccessibility;
exports.getRenderedAttribute = handlesAccessibility_1.getRenderedAttribute;
var isConformant_1 = require("./isConformant");
exports.isConformant = isConformant_1.isConformant;
var htmlIsAccessibilityCompliant_1 = require("./htmlIsAccessibilityCompliant");
exports.htmlIsAccessibilityCompliant = htmlIsAccessibilityCompliant_1.htmlIsAccessibilityCompliant;
