// @pnp/spfx-property-controls/lib/propertyFields/teamPicker/PropertyFieldTeamPickerHost.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import * as React from "react";
import { Label } from "office-ui-fabric-react/lib/Label";
import { SearchBox } from "office-ui-fabric-react/lib/SearchBox";
import { Spinner, SpinnerSize } from "office-ui-fabric-react/lib/Spinner";
import * as strings from "PropertyControlStrings";
import { initializeIcons } from "@uifabric/icons";
import { Async } from "@uifabric/utilities/lib";
import * as telemetry from "../../common/telemetry";
import TeamsSearchService from "../../services/TeamsSearchService";
import FieldErrorMessage from "../errorMessage/FieldErrorMessage";
import styles from "./PropertyFieldTeamPickerHost.module.scss";
import { PropertyFieldTeamPickerListItem } from "./PropertyFieldTeamPickerListItem/PropertyFieldTeamPickerListItem";
var PropertyFieldTeamPickerHost = /** @class */ (function (_super) {
    __extends(PropertyFieldTeamPickerHost, _super);
    function PropertyFieldTeamPickerHost(props) {
        var _this = _super.call(this, props) || this;
        _this.onSearchFieldChange = function (newValue) { return __awaiter(_this, void 0, void 0, function () {
            var teams, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(newValue && newValue.length > 2)) return [3 /*break*/, 6];
                        this.setState({ isLoading: true });
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        return [4 /*yield*/, this.teamsService.searchTeams(this.props.context, newValue)];
                    case 2:
                        teams = _a.sent();
                        this.setState({ teamSearchResults: teams });
                        return [3 /*break*/, 5];
                    case 3:
                        error_1 = _a.sent();
                        this.setState({ errorMessage: error_1.message });
                        return [3 /*break*/, 5];
                    case 4:
                        this.setState({ isLoading: false });
                        return [7 /*endfinally*/];
                    case 5: return [3 /*break*/, 7];
                    case 6:
                        this.setState({ teamSearchResults: [] });
                        _a.label = 7;
                    case 7: return [2 /*return*/];
                }
            });
        }); };
        _this.handleCheckboxChange = function (team, checked) {
            var selectedTeams = _this.state.selectedTeams.slice();
            if (checked) {
                if (_this.props.multiSelect) {
                    selectedTeams.push(team);
                }
                else {
                    selectedTeams = [team];
                }
            }
            else {
                if (_this.props.multiSelect) {
                    selectedTeams.splice(selectedTeams.indexOf(team), 1);
                }
                else {
                    selectedTeams = [];
                }
            }
            _this.props.onPropertyChange(_this.props.targetProperty, _this.state.selectedTeams, selectedTeams);
            // Trigger the apply button
            if (typeof _this.props.onChange !== 'undefined' && _this.props.onChange !== null) {
                _this.props.onChange(_this.props.targetProperty, selectedTeams);
            }
            _this.setState({ selectedTeams: selectedTeams });
        };
        initializeIcons();
        telemetry.track('PropertyFieldTeamPicker', {
            disabled: props.disabled
        });
        _this.state = {
            isLoading: false,
            selectedTeams: props.initialTeams || [],
            teamSearchResults: [],
            errorMessage: null
        };
        _this.async = new Async(_this);
        _this.teamsService = new TeamsSearchService();
        return _this;
    }
    PropertyFieldTeamPickerHost.prototype.render = function () {
        var _this = this;
        var _a = this.state, isLoading = _a.isLoading, teamSearchResults = _a.teamSearchResults, selectedTeams = _a.selectedTeams;
        return (React.createElement("div", null,
            this.props.label && React.createElement(Label, null, this.props.label),
            React.createElement(SearchBox, { placeholder: strings.TeamPickerSearchBoxPlaceholder, onChanged: this.async.debounce(this.onSearchFieldChange, this.props.deferredValidationTime) }),
            isLoading &&
                React.createElement(Spinner, { size: SpinnerSize.medium }),
            !isLoading && teamSearchResults &&
                React.createElement("div", null,
                    teamSearchResults.length > 0 &&
                        React.createElement("ul", { className: styles.siteList }, teamSearchResults.map(function (team) {
                            return React.createElement(PropertyFieldTeamPickerListItem, { key: team.id, checked: selectedTeams.filter(function (s) { return s.id === team.id; }).length > 0, handleCheckboxChange: _this.handleCheckboxChange, team: team });
                        })),
                    teamSearchResults.length === 0 &&
                        React.createElement(Label, null, strings.TeamPickerNoResults)),
            selectedTeams && selectedTeams.length > 0 &&
                React.createElement("div", null,
                    React.createElement(Label, { className: styles.bold },
                        selectedTeams.length,
                        " ",
                        strings.TeamPickerSitesChosen),
                    React.createElement("ul", { className: styles.siteList }, selectedTeams.map(function (team) {
                        return React.createElement(PropertyFieldTeamPickerListItem, { key: team.id, checked: selectedTeams.filter(function (s) { return s.id === team.id; }).length > 0, handleCheckboxChange: _this.handleCheckboxChange, team: team });
                    }))),
            React.createElement(FieldErrorMessage, { errorMessage: this.state.errorMessage })));
    };
    return PropertyFieldTeamPickerHost;
}(React.Component));
export default PropertyFieldTeamPickerHost;
