// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Menu/menuItemIndicatorStyles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var utils_1 = require("../../../../utils");
var submenuIndicatorUrl_1 = require("./submenuIndicatorUrl");
exports.menuItemIndicatorStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables, rtl = _a.rtl;
        return tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({ position: 'relative', float: 'right', left: utils_1.pxToRem(12), userSelect: 'none', marginRight: utils_1.pxToRem(4) }, (p.inSubmenu && {
            position: 'absolute',
            top: utils_1.pxToRem(6),
            right: utils_1.pxToRem(2),
            left: 'unset',
        })), (rtl && {
            transform: "scaleX(-1)",
        })), { content: '" "', display: 'block', overflow: 'hidden', height: utils_1.pxToRem(16), width: utils_1.pxToRem(16), backgroundSize: utils_1.pxToRem(16), backgroundImage: submenuIndicatorUrl_1.submenuIndicatorUrl(v.indicatorColor, p.vertical) }), (p.active && tslib_1.__assign({ backgroundImage: submenuIndicatorUrl_1.submenuIndicatorUrl(v.activeIndicatorColor, p.vertical) }, (p.primary && tslib_1.__assign({ backgroundImage: submenuIndicatorUrl_1.submenuIndicatorUrl(v.activePrimaryIndicatorColor, p.vertical) }, (p.vertical && {
            backgroundImage: submenuIndicatorUrl_1.submenuIndicatorUrl(v.activePrimaryVerticalIndicatorColor, p.vertical),
        })))))), (p.underlined && { backgroundImage: submenuIndicatorUrl_1.submenuIndicatorUrl(v.indicatorColor, p.vertical) })), (p.iconOnly && { backgroundImage: submenuIndicatorUrl_1.submenuIndicatorUrl(v.indicatorColor, p.vertical) }));
    },
};
