// @pnp/spfx-controls-react/lib/controls/chartControl/ChartControl.module.scss.js
/* tslint:disable */
require("./ChartControl.module.css");
var styles = {
    defaultFontColor: '#666666',
    defaultFontFamily: '"Segoe UI", -apple-system, BlinkMacSystemFont, "Roboto", "Helvetica Neue", sans-serif',
    defaultFontSize: '14px',
    lineColor: '#f4f4f4',
    titleColor: '#666666',
    titleFont: '"Segoe UI", -apple-system, BlinkMacSystemFont, "Roboto", "Helvetica Neue", sans-serif',
    titleFontSize: '14px',
    legendColor: '#a6a6a6',
    legendFont: '"Segoe UI", -apple-system, BlinkMacSystemFont, "Roboto", "Helvetica Neue", sans-serif',
    legendFontSize: '12px',
    tooltipFontSize: '12px',
    tooltipFont: '"Segoe UI", -apple-system, BlinkMacSystemFont, "Roboto", "Helvetica Neue", sans-serif',
    tooltipBodyColor: '#ffffff',
    tooltipTitleFont: '"Segoe UI", -apple-system, BlinkMacSystemFont, "Roboto", "Helvetica Neue", sans-serif',
    tooltipTitleColor: '#ffffff',
    tooltipTitleFontSize: '12px',
    tooltipFooterColor: '#ffffff',
    tooltipFooterFont: '"Segoe UI", -apple-system, BlinkMacSystemFont, "Roboto", "Helvetica Neue", sans-serif',
    tooltipFooterFontSize: '12px',
    tooltipBackgroundColor: '#333333',
    tooltipBorderColor: '#333333',
    scaleFontColor: '#666666',
    scaleFont: '"Segoe UI", -apple-system, BlinkMacSystemFont, "Roboto", "Helvetica Neue", sans-serif',
    scaleFontSize: '12px',
    chartComponent: 'chartComponent_7c0259eb',
    themed: 'themed_7c0259eb',
    accessibleTable: 'accessibleTable_7c0259eb'
};
export default styles;
/* tslint:enable */
