// @fluentui/react-northstar/dist/commonjs/components/Menu/MenuItemContent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.MenuItemContent = exports.menuItemContentClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var menuItemContentClassName = 'ui-menu__itemcontent';
/**
 * A MenuItemContent allows a user to have a dedicated component that can be targeted from the theme.
 */

exports.menuItemContentClassName = menuItemContentClassName;
var MenuItemContent = (0, _reactBindings.compose)(_Box.Box, {
  className: menuItemContentClassName,
  displayName: 'MenuItemContent',
  mapPropsToStylesProps: function mapPropsToStylesProps(props) {
    return {
      hasMenu: props.hasMenu,
      hasIcon: props.hasIcon,
      vertical: props.vertical,
      inSubmenu: props.inSubmenu
    };
  },
  handledProps: ['hasMenu', 'hasIcon', 'vertical', 'inSubmenu'],
  overrideStyles: true,
  shorthandConfig: {
    mappedProp: 'content'
  }
});
exports.MenuItemContent = MenuItemContent;
MenuItemContent.defaultProps = {
  as: 'span'
};
MenuItemContent.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon(), {
  hasIcon: PropTypes.bool,
  hasMenu: PropTypes.bool,
  vertical: PropTypes.bool,
  inSubmenu: PropTypes.bool
});
