// @pnp/spfx-property-controls/lib/propertyFields/webPartInformation/PropertyPaneWebPartInformation.js
import * as React from 'react';
import * as ReactDom from 'react-dom';
import { PropertyPaneFieldType } from '@microsoft/sp-webpart-base';
import PropertyPaneWebPartInformationHost from './PropertyPaneWebPartInformationHost';
var PropertyPaneWebPartInformationBuilder = /** @class */ (function () {
    function PropertyPaneWebPartInformationBuilder(_properties) {
        this.type = PropertyPaneFieldType.Custom;
        this.properties = {
            key: _properties.key,
            moreInfoLink: _properties.moreInfoLink,
            moreInfoLinkTarget: _properties.moreInfoLinkTarget !== undefined ? _properties.moreInfoLinkTarget : "_blank",
            videoProperties: _properties.videoProperties,
            description: _properties.description,
            onRender: this.onRender.bind(this)
        };
    }
    PropertyPaneWebPartInformationBuilder.prototype.render = function () {
        if (!this.elem) {
            return;
        }
        this.onRender(this.elem);
    };
    PropertyPaneWebPartInformationBuilder.prototype.onRender = function (elem, ctx, changeCallback) {
        if (!this.elem) {
            this.elem = elem;
        }
        var element = React.createElement(PropertyPaneWebPartInformationHost, {
            moreInfoLink: this.properties.moreInfoLink,
            moreInfoLinkTarget: this.properties.moreInfoLinkTarget,
            description: this.properties.description,
            videoProperties: this.properties.videoProperties
        });
        ReactDom.render(element, elem);
    };
    return PropertyPaneWebPartInformationBuilder;
}());
export function PropertyPaneWebPartInformation(properties) {
    return new PropertyPaneWebPartInformationBuilder(properties);
}
