// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Input/inputVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../../../utils");
exports.inputVariables = function (siteVars) { return ({
    colorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    colorError: siteVars.colorScheme.red.foreground,
    borderColorError: siteVars.colorScheme.red.background,
    iconPosition: 'absolute',
    iconRight: utils_1.pxToRem(10),
    iconLeft: utils_1.pxToRem(6),
    inputPaddingWithIconAtStart: utils_1.pxToRem(5) + " " + utils_1.pxToRem(12) + " " + utils_1.pxToRem(5) + " " + utils_1.pxToRem(24),
    inputPaddingWithIconAtEnd: utils_1.pxToRem(5) + " " + utils_1.pxToRem(24) + " " + utils_1.pxToRem(5) + " " + utils_1.pxToRem(12),
    inputPadding: utils_1.pxToRem(5) + " " + utils_1.pxToRem(12),
    inputInsideLabelPaddingTop: utils_1.pxToRem(14),
    borderColor: 'transparent',
    borderRadius: utils_1.pxToRem(3) + " " + utils_1.pxToRem(3) + " " + utils_1.pxToRem(2) + " " + utils_1.pxToRem(2),
    borderWidth: "0 0 " + utils_1.pxToRem(2) + " 0",
    backgroundColor: siteVars.colorScheme.default.background2,
    backgroundColorInverted: siteVars.colorScheme.default.background,
    fontColor: siteVars.colorScheme.default.foreground,
    fontSize: siteVars.fontSizes.medium,
    iconColor: siteVars.colorScheme.default.foreground,
    successfulColor: siteVars.colorScheme.green.foreground,
    inputFocusBorderColor: "transparent transparent " + siteVars.colorScheme.brand.borderFocus1 + " transparent",
    inputFocusBorderRadius: utils_1.pxToRem(3) + " " + utils_1.pxToRem(3) + " " + utils_1.pxToRem(2) + " " + utils_1.pxToRem(2),
    placeholderColor: siteVars.colorScheme.default.foreground1,
}); };
