// @fluentui/styles/dist/dts/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var callable_1 = require("./callable");
exports.callable = callable_1.callable;
var createTheme_1 = require("./createTheme");
exports.createTheme = createTheme_1.createTheme;
var debugEnabled_1 = require("./debugEnabled");
exports.isDebugEnabled = debugEnabled_1.isEnabled;
var deepmerge_1 = require("./deepmerge");
exports.deepmerge = deepmerge_1.deepmerge;
var mergeThemes_1 = require("./mergeThemes");
exports.mergeThemes = mergeThemes_1.mergeThemes;
exports.emptyTheme = mergeThemes_1.emptyTheme;
exports.mergeComponentVariables = mergeThemes_1.mergeComponentVariables;
exports.mergeComponentStyles = mergeThemes_1.mergeComponentStyles;
exports.mergeFontFaces = mergeThemes_1.mergeFontFaces;
exports.mergeStaticStyles = mergeThemes_1.mergeStaticStyles;
exports.mergeStyles = mergeThemes_1.mergeStyles;
exports.mergeThemeStyles = mergeThemes_1.mergeThemeStyles;
exports.mergeThemeVariables = mergeThemes_1.mergeThemeVariables;
exports.mergeSiteVariables = mergeThemes_1.mergeSiteVariables;
var objectKeysToValues_1 = require("./objectKeysToValues");
exports.objectKeyToValues = objectKeysToValues_1.objectKeyToValues;
var withDebugId_1 = require("./withDebugId");
exports.withDebugId = withDebugId_1.withDebugId;
tslib_1.__exportStar(require("./types"), exports);
