// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Carousel/carouselVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../../../utils");
exports.carouselVariables = function (siteVars) { return ({
    width: 300,
    height: 300,
    paddlePreviousSize: 32,
    paddleNextSize: 32,
    focusOuterBorderColor: siteVars.focusOuterBorderColor,
    focusOuterBorderRadius: siteVars.borderRadius,
    paddleHeight: utils_1.pxToRem(32),
    paddleColor: siteVars.colorScheme.default.foreground,
    paddleBackgroundColor: siteVars.colorScheme.default.background,
    paddleBorderRadius: siteVars.borderRadius,
    paddleBorderColor: siteVars.colorScheme.default.border,
    paddleBoxShadow: siteVars.shadowLevel1,
    paddleColorHover: siteVars.colorScheme.brand.foregroundHover,
    paddleBackgroundColorHover: siteVars.colorScheme.default.backgroundHover2,
    paddleBorderColorHover: siteVars.colorScheme.default.borderHover,
    paddleColorActive: siteVars.colorScheme.default.foregroundPressed,
    paddleBackgroundColorActive: siteVars.colorScheme.default.backgroundPressed,
    paddleBorderColorActive: siteVars.colorScheme.default.borderPressed,
    paddleBackgroundColorFocus: undefined,
    paddleBorderColorFocus: undefined,
    paddleColorFocus: undefined,
    paddleColorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    paddleBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
    paddleBorderColorDisabled: 'transparent',
    paddleIndicatorSize: utils_1.pxToRem(16),
    focusBorderRadius: siteVars.borderRadius,
    focusBorderWidth: siteVars.borderWidth,
    focusInnerBorderColor: siteVars.focusInnerBorderColor,
    focusBorderZIndex: siteVars.zIndexes.foreground,
}); };
